c++/src/kj/string.h

```diff
@@ -451,6 +548,53 @@ auto toCharSequence(T&& value) -> decltype(_::STR * kj::fwd<T>(value)) {
 }
 
 namespace _ {  // private
+template <typename T, typename = void>
+struct IsStringifiedToSignalSafeCharSequence_ {
+  // This is the base implementation for non-stringifiable types (KJ_STRINGIFY not implemented).
+  static constexpr bool value = false;
+};
+template <typename T>
+struct IsStringifiedToSignalSafeCharSequence_<T, VoidSfinae<decltype(STR * instance<T>())>> {
+  // This specialization restricts itself to instances of T that are stringifiable.
+
+  static constexpr bool value = isSpecialization<
+      decltype(STR * instance<T>()), SignalSafeCharSequence>();
+  // This is true when the result of stringifying T results in a SignalSafeCharSequence.
+};
+
+template <typename T>
+struct IsSignalSafeToCharSequence_ {
+  static constexpr bool value = IsStringifiedToSignalSafeCharSequence_<T>::value;
+  // This double-dispatch is needed so that ArrayPtr<T> works correctly. Otherwise the SFINAE
+  // check for T being stringifiable would be ambiguous in some cases due to the implicit handling
+  // of wrapped array types below.
+};
+template <typename T>
+struct IsSignalSafeToCharSequence_<SignalSafeCharSequence<T>> {
+  static constexpr bool value = true;
+};
+// Rather than needing to manually annotate these, handle collections automatically.
```

Comment:
I don't like that these declarations are separated from the stringification implementations they are describing. I especially don't like it because these declarations are inconsistent with said stringification implementations, which may have been more obvious if the stringification implementations themselves took care of declaring their own signal-safety.

In particular:

- `KJ_STRINGIFY(Delimited<T>&& delimited)` is later declared to return `SignalSafeCharSequence`, even though it's only signal-safe if `T` can be stringified signal-safely.
- `Array<T>` and `ArrayPtr<T>` delegate to `Delimited<T>`.
- AFAICT there are no stringifiers for `CappedArray<T>` and `FixedArray<T>` except when `T` is `char`. It seems weird to declare signal-safety for implementations that don't even exist.

Instead of specializing all these types here, how about declaring a type `PropagateSignalSafeStringifiable<T, U>` which evaluates to `SignalSafeCharSequence<T>` if `U` is signal-safe-stringifiable, or otherwise evaluates to just `T`.

Then you can declare the `Delimited` stringifier like:

```
template <typename T>
PropagateSignalSafeStringifiable<Delimited<T>, T> KJ_STRINGIFY(Delimited<T>&& delimited);
```

And similarly for Array/ArrayPtr.